### PR TITLE
Launch immersive experiences directly

### PR DESCRIPTION
**Pull request by felipeerias**
This PR implements support for opening immersive experiences directly as soon as the application is launched.

We do this by adding a built-in extension which will activate a particular element in the page.

We need to set the preference `dom.vr.require-gesture` to false so this script can launch immersive WebXR experiences without user input. Media autoplay also need to be allowed in this mode.

The information required to identify the element to activate is passed as parameters to the Intent:

- `launch_immersive`
- `launch_immersive_parent_xpath`
- `launch_immersive_element_xpath`
- a target URL to open

One additional concern is that we should only change the configuration prefs when that is needed for launching Wolvic directly in immersive mode.

The problem is that we only write that prefs file the first time that we create the browser engine, which happens several times at launch time **before** we have parsed the incoming Intent and we can know in which mode we are launching.

The solution is that `WidgetManagerDelegate.isOpenInImmersive()` internally will check the current Intent, to cover the case where it is being called too early.

We then pass `VRBrowserActivity` as the context to the objects that may initialize the engine, so eventually `SessionUtils.prepareConfigurationPath()` is able to decide whether specific prefs should be added to the configuration file.

For launching Wolvic directly in immersive mode, these prefs are:

```
  dom.vr.require-gesture: false
  media.autoplay.default: 0
```

### CODE REVIEW

**Comment by felipeerias** [comment]
This PR can be tested from the command line with:

```shell
adb shell am start -n "com.igalia.wolvic/com.igalia.wolvic.VRBrowserActivity" \
    -a android.intent.action.MAIN                                             \
    -c android.intent.category.LAUNCHER                                       \
    -a android.intent.action.VIEW                                             \
    -d "https://heyvr.io/arcade/games/barista-express"                        \
    --ez open_in_immersive true                                               \
    --ez hide_webxr_interstitial true                                         \
    -e open_in_immersive_parent_id game-frame                                 \
    -e open_in_immersive_element_xpath '/html/body/a-scene/div[2]/button'
```

**Comment by felipeerias** [comment]
I have updated the PR so that it uses the XPath of the parent element, which is a more flexible solution.

**Review by svillar — COMMENTED** [review]
I think I told @felipeerias about this in private, but I'll add it here for everybody interested to know. The idea of using an extension is nice, but the problem is that the Chromium backend won't have extensions support initially as it's a lot of work. So provided that we don't have a solution at the moment I'd prioritize one that does not involve extensions.

That said I agree that having an extension that inspects the DOM is likely the best we could do, so I indeed have mixed feelings.

**Comment by felipeerias** [comment]
This implementation might be the best that we can do, but it still has several issues.

First of all, the command to launch Wolvic in immersive mode needs to know the exact element in the page which will open the WebXR experience. If the page changes, it will be necessary to update that command.

Secondly, the approach of using a Web extension to trigger specific elements in the page will not be possible on Chromium, because we don't support extensions there just yet.

Third, in some cases our extension is not able to reach the immersive button because the WebXR experience is inside of an iframe from a different origin, and security rules prevent us from accessing a cross-origin object from the embedding page. It might be possible to work around this issue by passing messages between the content scripts running on the page and on the iframe.

This last problem is particularly painful because it affects HeyVR, a large repository of WebXR games: the pages are in `https://heyvr.io` but their iframes point towards `https://games.heyvr.io`.

Finally, some pages might have specific quirks that we have to consider. For example, Moon Rider requires two clicks to open the immersive experience, and Magical Reflections requires the user to accept cookies before anything else.

With all that in mind, the implementation in this PR does work reasonably well. Check out the following video:

https://github.com/Igalia/wolvic/assets/1412060/5a1bce40-b91d-4dd9-936f-d5bc2f3c2221

Command to launch A-Painter:

```
adb shell am start -n "com.igalia.wolvic.dev/com.igalia.wolvic.VRBrowserActivity" \
    -a android.intent.action.VIEW                                                 \
    -c android.intent.category.LAUNCHER                                           \
    -d "https://aframe.io/a-painter/"                                             \
    --ez open_in_immersive true                                                   \
    --ez hide_webxr_interstitial true                                             \
    -e open_in_immersive_element_xpath '\/html\/body\/a-scene\/div[2]\/button'
```

Command to launch Moon Rider:

```
adb shell am start -n "com.igalia.wolvic.dev/com.igalia.wolvic.VRBrowserActivity" \
    -a android.intent.action.VIEW                                                 \
    -c android.intent.category.LAUNCHER                                           \
    -d "https://moonrider.xyz/"                                                   \
    --ez open_in_immersive true                                                   \
    --ez hide_webxr_interstitial true                                             \
    -e open_in_immersive_element_xpath '\/\/\*[@id=\"vrButton\"]'
```

File: app/src/main/assets/extensions/wolvic_autowebxr/content.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    // The parent element is typically an iframe and, if it comes from a different origin,
+    // we might not be able to access its contents directly.
+    // If parentElementXPath is null, we will use the root to find the target element.
+
+    var parent, parentDocument;
+    if (parentElementXPath) {
+        parent = getElementByXPath(document, parentElementXPath);
+        if (!parent) {
+            logDebug('Parent element not found, retrying');
+            setTimeout(clickImmersiveElement, 1000);
```
Comment (svillar) [inline app/src/main/assets/extensions/wolvic_autowebxr/content.js:48]: This has the potential of trigger a never ending sequence of timeouts if the parent element is not correctly specified. I guess you added the timeout to allow the page to load, but in that case I think it's safer to listen to DOMContentLoaded or similar event.

File: app/src/main/assets/extensions/wolvic_autowebxr/content.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    } else {
+        logDebug('Parent is loaded');
+    }
+
+    let targetElement = getElementByXPath(parentDocument, targetElementXPath);
+
+    if (targetElement) {
+        logDebug('Target element found, calling click()');
+        targetElement.click();
+        targetElement.click();
```
Comment (svillar) [inline app/src/main/assets/extensions/wolvic_autowebxr/content.js:97]: Why two?

File: app/src/common/shared/com/igalia/wolvic/VRBrowserActivity.java [hunk 3]
```diff
@@ -128,6 +128,12 @@ public class VRBrowserActivity extends PlatformActivity implements WidgetManager
     public static final String EXTRA_KIOSK = "kiosk";
     private static final long BATTERY_UPDATE_INTERVAL = 60 * 1_000_000_000L; // 60 seconds
 
+    boolean mOpenInImmersive = false;
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/VRBrowserActivity.java:131]: I think this should be private

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java [hunk 4]
```diff
@@ -1470,6 +1473,38 @@ public void openInKioskMode(@NonNull String aUri) {
         mFocusedWindow.setKioskMode(true);
     }
 
+    public void openInImmersiveMode(Uri targetUri, String immersiveParentElementXPath, String immersiveTargetElementXPath) {
+        String extensionId = "wolvic-autowebxr@igalia.com";
+        String extensionUrl = "resource://android/assets/extensions/wolvic_autowebxr/";
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java:1478]: Can we move these two to the top of the file with the other declarations

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java [hunk 5]
```diff
@@ -68,6 +69,8 @@ public class Windows implements TrayListener, TopBarWidget.Delegate, TitleBarWid
 
     public static final int WHITE = 0xFFFFFFFF;
     public static final int GRAY = 0x555555FF;
+    public static final String PARENT_ELEMENT_XPATH_PARAMETER = "wolvic-autowebxr-parentElementXPath";
+    public static final String TARGET_ELEMENT_XPATH_PARAMETER = "wolvic-autowebxr-targetElementXPath";
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java:73]: private?

**Review by svillar — CHANGES_REQUESTED** [review]

File: app/src/main/assets/extensions/wolvic_autowebxr/content.js (hunk 2 above)
Comment (felipeerias) [reply app/src/main/assets/extensions/wolvic_autowebxr/content.js:97]: There was a case where one `click()` didn't work. I think that I will remove this and test again different experiences, to check if it is still needed.

File: app/src/common/shared/com/igalia/wolvic/VRBrowserActivity.java (hunk 3 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/VRBrowserActivity.java:131]: Yes, of course.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java (hunk 4 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java:1478]: Yes, good point, and they should be constants.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java (hunk 5 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/Windows.java:73]: Seems a better option, yes.

File: app/src/main/assets/extensions/wolvic_autowebxr/content.js (hunk 1 above)
Comment (felipeerias) [reply app/src/main/assets/extensions/wolvic_autowebxr/content.js:48]: Note that this code is called after document.readyState === 'complete' or the load event has been already received. The parent might still be unreachable because the website is still setting things up via JS.

My idea here was simply to call again clickImmersiveElement after a small delay. Maybe it makes sense to have a timer or a counter so we don't keep trying forever.

Comment (svillar) [reply app/src/main/assets/extensions/wolvic_autowebxr/content.js:48]: Ah right. In any case I think we need a counter there to prevent the potential never ending loop.

**Comment by felipeerias** [comment]
@svillar I have updated the PR following your comments. I also had to allow autoplay when we start in this mode to remove an error with Moon Rider, although for some reason it is still not working properly (it was working fine before).

Reference: https://wiki.mozilla.org/Media/block-autoplay

**Comment by felipeerias** [comment]
Access Mars:

```
 adb shell am start -n "com.igalia.wolvic/com.igalia.wolvic.VRBrowserActivity" \
    -a android.intent.action.VIEW                                                 \
    -c android.intent.category.LAUNCHER                                           \
    -d "https://accessmars.withgoogle.com/"                                                   \
    --ez open_in_immersive true                                                   \
    --ez hide_webxr_interstitial true                                             \
    -e open_in_immersive_element_xpath '\/html\/body\/div\[1\]\/header\/div\/button'
```

(Note that depending on your setup you might have to add `.dev` to the package name)

**Review by svillar — CHANGES_REQUESTED** [review]

File: app/src/common/shared/com/igalia/wolvic/VRBrowserActivity.java [hunk 6]
```diff
@@ -128,6 +128,12 @@ public class VRBrowserActivity extends PlatformActivity implements WidgetManager
     public static final String EXTRA_KIOSK = "kiosk";
     private static final long BATTERY_UPDATE_INTERVAL = 60 * 1_000_000_000L; // 60 seconds
 
+    private boolean mOpenInImmersive = false;
+    public static final String EXTRA_OPEN_IN_IMMERSIVE = "open_in_immersive";
+    // Element where a click would be simulated to launch the WebXR experience.
+    public static final String EXTRA_OPEN_IN_IMMERSIVE_PARENT_XPATH = "open_in_immersive_parent_xpath";
+    public static final String EXTRA_OPEN_IN_IMMERSIVE_ELEMENT_XPATH = "open_in_immersive_element_xpath";
```
[the remark this answers is not in the document]
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/VRBrowserActivity.java:135]: The other Intent parameters in that file are public.

I think it's a good idea for documentation purposes, as a way to make clear that these strings are supposed to be used outside of this class.

File: app/src/main/assets/extensions/wolvic_autowebxr/content.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+var targetElementXPath;
+
+function getElementByXPath(document, xpath) {
+    let result = document.evaluate(xpath, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null);
+    return result.singleNodeValue;
+}
+
+// Limit the number of times that we can try to launch the experience, to avoid an infinite loop.
+var retryCounter = 0;
+const RETRY_LIMIT = 20;
```
[the remark this answers is not in the document]
Comment (felipeerias) [reply app/src/main/assets/extensions/wolvic_autowebxr/content.js:21]: Yes, but that's a problem in general with this feature. Some experiences just take a long time to load from the network.

In this initial implementation, we show the browser window before launching the immersive experience. Eventually my plan is to go directly from the loading screen to the immersive experience.

File: app/src/main/res/raw/fxr_config.yaml [hunk 8]
```diff
@@ -26,4 +26,6 @@ prefs:
   browser.gesture.pinch.in: ''
   browser.gesture.pinch.out.shift: ''
   browser.gesture.pinch.in.shift: ''
-  apz.one_touch_pinch.enabled: false
+  apz.one_touch_pinch.enabled: false,
+  # allows scripts to open WebXR immersive sessions
+  dom.vr.require-gesture: false
```
[the remark this answers is not in the document]
Comment (felipeerias) [reply app/src/main/res/raw/fxr_config.yaml:31]: One possibility could be to do it in `SessionUtils.prepareConfigurationPath()`, which copies the configuration file from the assets to the device's filesystem. Basically, the idea would be to append an extra line to the file if needed:

```java
final String REQUIRE_GESTURE_CONFIG = "dom.vr.require-gesture: false";
WidgetManagerDelegate widgetManager = (WidgetManagerDelegate) aContext;

if (widgetManager.isOpenInImmersive()) {
        outputStream.write(REQUIRE_GESTURE_CONFIG.getBytes());
}
```

Another approach could be to have two files in the assets, and copy one or the other depending on the current configuration. Maybe that would be better?

[the remark this answers is not in the document]
Comment (svillar) [reply app/src/main/res/raw/fxr_config.yaml:31]: I think it's enough with the first approach you mentioned. Adding a line to the stream should make it

[the remark this answers is not in the document]
Comment (felipeerias) [reply app/src/main/res/raw/fxr_config.yaml:31]: I've been working on this but it is a bit more complex than I expected because `prepareConfigurationPath()` is called very early, when we have not parsed the incoming Intent yet and so we don't know that we will be in the automatically immersive mode.

Specifically, as a result of `VRBrowserActivity.onCreate()` calling

```java
`((VRBrowserApplication) getApplication()).onActivityCreate(this)`.
```

[the remark this answers is not in the document]
Comment (felipeerias) [reply app/src/main/res/raw/fxr_config.yaml:31]: I finally found a way to change the configuration prefs only when needed for launching in immersive mode.

The problem is that we only write the prefs file the first time that we create the browser engine, which happens several times at launch time **before** we have parsed the incoming Intent and we can know in which mode we are launching.

The solution is that `WidgetManagerDelegate.isOpenInImmersive()` internally will check the current Intent, to cover the case where it is being called too early.

We then pass `VRBrowserActivity` as the context to the objects that may initialize the engine, so eventually `SessionUtils.prepareConfigurationPath()` is able to decide whether specific prefs should be added to the configuration file.

For launching Wolvic directly in immersive mode, these prefs are:
```
  dom.vr.require-gesture: false
  media.autoplay.default: 0
```

**Comment by felipeerias** [comment]
I have thoroughly updated this PR.

Unfortunately, it is hard to find experiences which will work well with it:

- some will show an ad before the game, like [heyvr.io](https://heyvr.io/)
- some don't have stable XPath identifiers for the button that launches the experience, like [magische-spiegelungen.de](https://www.magische-spiegelungen.de/)
- some work from the JS console via remote debugging but not from an extension (perhaps a timing issue?), like [moonrider.xyz](https://moonrider.xyz/)

These are a couple of experiences which work consistently.

A-Painter:

```
adb shell am start -n "com.igalia.wolvic.dev/com.igalia.wolvic.VRBrowserActivity" \
-a android.intent.action.VIEW                                                     \
-c android.intent.category.LAUNCHER                                               \
-d "https://aframe.io/a-painter/"                                                 \
--ez launch_immersive true                                                        \
--ez hide_webxr_interstitial true                                                 \
-e launch_immersive_element_xpath '\/html\/body\/a-scene\/div[2]\/button'
```

Access Mars:

```
adb shell am start -n "com.igalia.wolvic.dev/com.igalia.wolvic.VRBrowserActivity" \
-a android.intent.action.VIEW                                                     \
-c android.intent.category.LAUNCHER                                               \
-d "https://accessmars.withgoogle.com/"                                           \
--ez launch_immersive true                                                        \
--ez hide_webxr_interstitial true                                                 \
-e launch_immersive_element_xpath '\/html\/body\/div\[1\]\/header\/div\/button'
```

(remove the `.dev` from the package name if necessary)

File: app/src/common/shared/com/igalia/wolvic/VRBrowserApplication.java [hunk 9]
```diff
@@ -58,7 +58,7 @@ protected void onActivityCreate(@NonNull Context activityContext) {
         mConnectivityManager.init();
         mPlaces = new Places(activityContext);
         mServices = new Services(activityContext, mPlaces);
-        mLoginStorage = new LoginStorage(this);
+        mLoginStorage = new LoginStorage(activityContext);
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/VRBrowserApplication.java:61]: Hmm this seems unrelated although it looks like it's correct. Should we land it separatedely?

File: app/src/common/shared/com/igalia/wolvic/browser/LoginStorage.kt [hunk 10]
```diff
@@ -16,7 +16,7 @@ class LoginStorage(
         val context: Context
 ) {
 
-    private val places = (context as AppServicesProvider).places
+    private val places = (context.applicationContext as AppServicesProvider).places
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/browser/LoginStorage.kt:19]: Same comment

File: app/src/common/shared/com/igalia/wolvic/browser/PermissionDelegate.java [hunk 11]
```diff
@@ -221,7 +221,8 @@ public WResult<Integer> onContentPermissionRequest(WSession aSession, ContentPer
             // https://hacks.mozilla.org/2019/02/firefox-66-to-block-automatically-playing-audible-video-and-audio/
             return WResult.fromValue(ContentPermission.VALUE_ALLOW);
         } else if(perm.permission == PERMISSION_AUTOPLAY_AUDIBLE) {
-            if (SettingsStore.getInstance(mContext).isAutoplayEnabled()) {
+            // allow autoplay when we launch Wolvic in immersive mode automatically
+            if (SettingsStore.getInstance(mContext).isAutoplayEnabled() || mWidgetManager.isLaunchImmersive()) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/browser/PermissionDelegate.java:225]: Not sure I understand why this is needed. Why do you want to force autoplay for immersive?

File: app/src/common/shared/com/igalia/wolvic/browser/engine/SessionUtils.java [hunk 12]
```diff
@@ -18,24 +19,30 @@ class SessionUtils {
...
 
-        try (FileOutputStream outputStream = new FileOutputStream(path, false)) {
             int read;
             int bufferSize = 8192;
             byte[] bytes = new byte[bufferSize];
             while ((read = yaml.read(bytes)) != -1) {
                 outputStream.write(bytes, 0, read);
             }
+
+            if (aContext instanceof WidgetManagerDelegate && ((WidgetManagerDelegate) aContext).isLaunchImmersive()) {
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/browser/engine/SessionUtils.java:41]: Can aContext by anything different from WidgetManagerDelegate? If not I prefer we use an assert here and then the type casting directly.

File: app/src/main/AndroidManifest.xml [hunk 13]
```diff
@@ -43,7 +43,7 @@
         <activity
             android:name="com.igalia.wolvic.VRBrowserActivity"
             android:launchMode="singleInstance"
-            android:exported="false"
+            android:exported="true"
```
Comment (svillar) [inline app/src/main/AndroidManifest.xml:46]: I believe we cannot do this. The new main activity is the VersionCheck activity and is the only one that can be exported as far as I know. Note that VersionCheck copies all the parameters of the intent to the VRBrowserActivity so we don't really need to export this.

**Review by svillar — CHANGES_REQUESTED** [review]
Looking good, we're superclose to land this. I have some comments

File: app/src/common/shared/com/igalia/wolvic/VRBrowserApplication.java (hunk 9 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/VRBrowserApplication.java:61]: The creation of `LoginStorage` is the first time that `EngineProvider.getOrCreateRuntime()` is called on startup. We need to pass the activity Context so it can use the methods in `WidgetManagerDelegate`.

File: app/src/common/shared/com/igalia/wolvic/browser/PermissionDelegate.java (hunk 11 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/browser/PermissionDelegate.java:225]: Some experiences try to autoplay audio on launch.

File: app/src/common/shared/com/igalia/wolvic/browser/engine/SessionUtils.java (hunk 12 above)
Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/browser/engine/SessionUtils.java:41]: Yes, that was actually the case before this PR because `LoginStorage` was the first one to call `EngineProvider.getOrCreateRuntime()` with the application's `Context`.

The reason for not having an assert is that `getOrCreateRuntime()` is called from different places and in principle we don't know which one will trigger this codepath.

**Comment by felipeerias** [comment]
I had to modify `AndroidManifest.xml` so we could launch Wolvic with parameters without needing to specify the specific Activity.

Now that `VersionCheckActivity` is the entry point of our application and the only Activity that we export, it needs to use the intent filters that were previously associated to `VRBrowserActivity`.

These are the new comands to test this feature:

```
adb shell am start -p com.igalia.wolvic.dev \
-a android.intent.action.VIEW               \
-d "https://aframe.io/a-painter/"           \
--ez launch_immersive true                  \
--ez hide_webxr_interstitial true           \
-e launch_immersive_element_xpath '\/html\/body\/a-scene\/div[2]\/button'
```

```
adb shell am start -p com.igalia.wolvic.dev \
-a android.intent.action.VIEW               \
-d "https://accessmars.withgoogle.com/"     \
--ez launch_immersive true                  \
--ez hide_webxr_interstitial true           \
-e launch_immersive_element_xpath '\/html\/body\/div\[1\]\/header\/div\/button'
```

File: app/src/main/AndroidManifest.xml [hunk 14]
```diff
@@ -39,25 +39,19 @@
                 <category android:name="android.intent.category.LAUNCHER" />
                 <category android:name="com.oculus.intent.category.2D" />
             </intent-filter>
-        </activity>
-        <activity
-            android:name="com.igalia.wolvic.VRBrowserActivity"
-            android:launchMode="singleInstance"
-            android:exported="true"
-            android:configChanges="density|keyboard|keyboardHidden|navigation|orientation|screenSize|uiMode|locale|layoutDirection"
-            android:windowSoftInputMode="stateAlwaysHidden">
```
Comment (svillar) [inline app/src/main/AndroidManifest.xml:48]: Are you sure the filter for configuration changes should be moved to the version check? Does VRBrowserActivity inherit them. Because if not we're going to have problems.

File: app/src/common/shared/com/igalia/wolvic/VersionCheckActivity.java [hunk 15]
```diff
@@ -24,11 +17,16 @@ protected void onCreate(Bundle savedInstanceState) {
...
 
             // Start VRBrowserActivity if OS version is compatible
             Intent intent = new Intent(this, VRBrowserActivity.class);
-            if (extras != null)
+            if (extras != null) {
                 intent.putExtras(extras);
+            }
+            if (data != null) {
+                intent.setData(data);
+            }
```
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/VersionCheckActivity.java:29]: This looks like a regression of the VersionCheckActivity addition. Should we land it in a different PR?

Comment (felipeerias) [reply app/src/common/shared/com/igalia/wolvic/VersionCheckActivity.java]: Done in https://github.com/Igalia/wolvic/pull/1575

**Comment by felipeerias** [comment]
The change which assigned the intent filters to `VersionCheckActivity` has been already integrated by https://github.com/Igalia/wolvic/pull/1575

I have rebased this PR and it is once again ready for review.

**Review by svillar — APPROVED** [review]
OK let's land this
